crates/sui-core/src/state_accumulator.rs

```diff
@@ -458,71 +602,166 @@ impl StateAccumulator {
+            (_, Some((highest_acc, running_root))) => Some((highest_acc, running_root)),
+        };
+        Self {
+            store,
+            running_root_accumulator: Mutex::new(running_root_data),
+        }
     }
 
-    pub fn accumulate_live_object(acc: &mut Accumulator, live_object: &LiveObject) {
-        match live_object {
-            LiveObject::Normal(object) => {
-                acc.insert(object.compute_object_reference().2);
+    pub async fn accumulate_running_root(
+        &self,
+        epoch_store: &AuthorityPerEpochStore,
+        checkpoint_seq_num: CheckpointSequenceNumber,
+        checkpoint_acc: Option<Accumulator>,
+    ) -> SuiResult {
+        let _scope = monitored_scope("AccumulateRunningRootV2");
+        // first check if we need to wait for other checkpoints to be accumulated. This can happen for the end
+        // of epoch checkpoint. Create a new scope to avoid deadlocking while we wait
+        let await_checkpoints = {
+            match self.running_root_accumulator.lock().as_ref() {
+                Some((highest_accumulated, _)) if *highest_accumulated < checkpoint_seq_num - 1 => {
+                    // TODO: take the checkpoint and check for EndOfEpochData to confirm we're at
+                    // the end of the epoch
+                    true
+                }
+                Some(_) => false,
+                None => false,
             }
-            LiveObject::Wrapped(key) => {
-                acc.insert(
-                    bcs::to_bytes(&WrappedObject::new(key.0, key.1))
-                        .expect("Failed to serialize WrappedObject"),
+        };
+        if await_checkpoints {
+            debug!(
+                "Awaiting accumulation up to checkpoint {} for epoch {} before adding to running root",
+                checkpoint_seq_num - 1,
+                epoch_store.epoch()
+            );
+            epoch_store
+                .notify_read_running_root(checkpoint_seq_num - 1)
+                .await
+                .expect("Failed to notify read running root");
+        }
+
+        let mut guard = self.running_root_accumulator.lock();
+
+        if let Some((highest_accumulated, running_root)) = guard.as_mut() {
+            if checkpoint_seq_num == *highest_accumulated {
+                // Enforce idempotency
+                Ok(())
+            } else {
+                let checkpoint_acc = checkpoint_acc.unwrap_or_else(|| {
+                    epoch_store
+                        .get_state_hash_for_checkpoint(&checkpoint_seq_num)
+                        .expect("Failed to get checkpoint accumulator from disk")
+                        .expect("Expected checkpoint accumulator to exist")
+                });
+
+                running_root.union(&checkpoint_acc);
+                *highest_accumulated = checkpoint_seq_num;
+                epoch_store.insert_running_root_accumulator(&checkpoint_seq_num, running_root)?;
```

Comment:
i assume this error return would probably trigger a panic somewhere further up, but just in case, lets write the db before updating `highest_accumulated`

Reply:
actually, reading this again, we need to update both running_root and highest_accumulated, otherwise an early return would leave them in an inconsistent state

Reply:
Removed the in memory running root, so this edge case went away